### ISSUE TITLE
It would be great to add some samples how to use PicoEVB

### ISSUE THREAD

**Issue by hlide** [RHSResearchLLC/PicoEVB#4]
Now we have a Pico/NanoEVB, we may need some samples:

- How to read/write the host RAM through PCIe? suppose you want to make a guest [co]processor on the FPGA, you may want to reserve some host RAM for the guest [co]processor.
- How to share host RAM between guest (FPGA) and host (PC)?

**Comment by RHSResearchLLC** [comment]
Hello hlide,

The Xilinx DMA core isn't exactly about sharing memory per se- its about transferring data between the host and FPGA at high speed. The Xilinx sample project has a good working demonstration of this and runs on NanoEVB/PicoEVB.

PCIe isn't really about shared memory, its about memory transfer. It seems shared because multiple clients can "access" it- but that access always consists of a transfer across the PCIe bus.

In your example (2), if the FPGA wants access to system RAM, it has to ask the PCIe system to transfer over a certain amount of data at a certain address. Likewise the FPGA can write into system RAM, but it still has to send a packet on the PCIe bus to do so. The FPGA has access to the RAM but it isn't like its a chip connected to the FPGA- you still need some real RAM in the FPGA to use as a reservoir.

The Xilinx DMA example demonstrates this: 

- The host allocates some ram, and fills it with some data pattern.
- The host sends this data to the FPGA via PCIe; the data is transferred the data directly into the FPGA's BRAM
- The Host requests some data from the FPGA
- The FPGA transfers data from BRAM directly into host memory via PCIe

Granted the Xilinx example isn't super easy to follow. I could make a simpler, more explicit example.

**Comment by hlide** [comment]
I know what DMA is and cannot be viewed as a direct RAM access. I asked you whether it was possible for the FPGA directly to address host RAM through PCIe before I decided to buy it (was my condition) and was kind of answered positively. But from what I can read above, there is no such possibility and a virtual RAM (a cache) is probably needed so the processor may not be impacted so much by the transfer timing between FPGA and Host through PCIe.

**Comment by RHSResearchLLC** [comment]
Hello hlide,

It seems maybe I don't understand exactly what you mean. The FPGA does have
access to host memory. This access is implemented in Xilinx PCIe endpoint.
Xilinx provides an AXI interface to host memory. The sample project simply
transfers data from host memory to block ram. But you can make your own AXI
master to interface to Xilinx endpoint instead.

If you feel like you were deceived, or if this board can't do what you
need, shoot me an email (dave@nanoevb.com) so we can work it out. If you
want a refund, I will make good on it.

—
You are receiving this because you commented.
Reply to this email directly, view it on GitHub
<https://github.com/RHSResearchLLC/PicoEVB/issues/4#issuecomment-357039153>,
or mute the thread

**Comment by hlide** [comment]
Suppose I want to add a soft processor with a virtual SRAM on the FPGA and that SRAM storage is provided by the host memory. So it is possible but you need a component which mimics an SRAM interface which is accessing the host memory through an AXI interface to host memory, right? So the issue is more with finding the right samples to show how to handle Xilinx endpoint to do that.

What I fear is there are no various samples which show how to program picoEMV for different tasks and so the product may remain unused. I had bought an MYIR Z-TURN ZYNQ 7020 long time ago and it remains in its box because there are no practical samples to learn how to use it efficiently in the end and I know several people in the same case with this board.

**Comment by RHSResearchLLC** [comment]
Yes you are correct, you would need a component that has AXI on one side and SRAM-like interface on the other (with some type of wait signal to coordinate with the receipt of the data from the PCIe bus). I don't know of any samples out there that cover this.

**Comment by hlide** [comment]
That's just an example.  I wonder if I can find a solution here: https://www.xilinx.com/products/intellectual-property/axi_pcie.html.
